server/src/main/java/org/elasticsearch/index/mapper/AnnotatedTextFieldMapper.java

```diff
@@ -0,0 +1,815 @@
+            }   
+        }
+        
+        @Override
+        public String toString() {
+           StringBuilder sb = new StringBuilder();
+           sb.append(textMinusMarkup);
+           sb.append("\n");
+           annotations.forEach(a -> {sb.append(a); sb.append("\n");});
+           return sb.toString();
+        }
+
+        public int numAnnotations() {
+            return annotations.size();
+        }
+
+        public AnnotationToken getAnnotation(int index) {
+            return annotations.get(index);
+        }   
+    }
+    
+    public static final class AnnotationAnalyzerWrapper extends AnalyzerWrapper {
+        
+
+        private final Analyzer delegate;
+
+        public AnnotationAnalyzerWrapper (Analyzer delegate) {
+          super(delegate.getReuseStrategy());
+          this.delegate = delegate;
+        }
+
+        /**
+         * Wraps {@link StandardAnalyzer}. 
+         */
+        public AnnotationAnalyzerWrapper() {
+          this(new StandardAnalyzer());
+        }
+        
+
+        @Override
+        public Analyzer getWrappedAnalyzer(String fieldName) {
+          return delegate;
+        }     
+
+        @Override
+        protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComponents components) {
+            if(components instanceof AnnotatedTokenStreamComponents){
+                // already wrapped. TODO not fully sure if this is the right thing to do here?
+                return components;
+            }
+            AnnotationsFilter filter = new AnnotationsFilter(components.getTokenStream());
+            return new AnnotatedTokenStreamComponents(components.getTokenizer(), filter);
+        }
+      }
+    
+    public static final class AnnotatedTokenStreamComponents extends TokenStreamComponents{
+
+        private AnnotationsFilter annotationsFilter;
+
+        public AnnotatedTokenStreamComponents(Tokenizer source, AnnotationsFilter annotationsFilter) {
+            super(source, annotationsFilter);
+            this.annotationsFilter = annotationsFilter;
+        }
+
+        @Override
+        protected void setReader(Reader reader) {
+            
+            AnnotatedText parsedText;
+            try {
+                parsedText = new AnnotatedText(readToString(reader));
+            } catch (IOException e) {
+                //TODO better exception
+                throw new RuntimeException(e);
+            }
+            annotationsFilter.setAnnotations(parsedText);
+ 
+            super.setReader(new StringReader(parsedText.textMinusMarkup));            
+        }
+        
+
+        private static String readToString(Reader reader) throws IOException{
+            //TODO FIXME I think this loses any end-of-text markers that are used
+            // internally in Lucene to turn arrays of strings into single strings.
+            // I think the positionIncrementGap logic looks for these special markers
```

Comment:
The logic for the position increment gap is not based on markers, multi-valued fields are indexed in multiple fields that share the same field name, the index writer registers the last position for each field and applies the posInc gap to the last position seen for the field being processed. Is this comment related to highlighting ? We don't apply the posInc gap during highlighting because positions are not important (highlighting is based only on offsets).

Reply:
I'm currently working on an issue with highlighting on fields that have arrays of text rather than a single value. There's all sorts of "magic" going on in unified highighter to assemble and disassemble strings to and from arrays using special marker characters.

Reply:
The `unified` highlighter uses `U+0000` to separate the values but it retrieves the original text first so this should not clash with the annotations? What is the issue with multi-values ? Can you explain the problem you're seeing ?

Reply:
Generally invalid offsets. I need to give the highlighter both marked-up and non-marked-up versions of the text. Marked-up text is required for any re-analysis to get the annotations but plain text is used for BreakIterator and PassageFormatter. I think the confusion comes when `content.length()` is used to denote end of a passage and the highlighter is using the marked-up version so is too long. I had code to reset the last passage's endOffset properly but that doesn't work properly in a multi-value because there's multiple passages with the wrong offsets. Maybe the better approach is to smuggle the marked-up version to the choice of analyzer and keep everything else in the highlighter (PassageFormatters, BreakIterators etc) working with just the plain-text.